File: backend-php/tests/TestRunner.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../src/Calculator.php';

function assert_equals(mixed $expected, mixed $actual, string $label): void
{
    if ($expected !== $actual) {
        throw new RuntimeException($label . ' expected ' . var_export($expected, true) . ' but got ' . var_export($actual, true));
    }
}

function assert_float_equals(float $expected, float $actual, string $label): void
{
    if (abs($expected - $actual) > 0.000001) {
        throw new RuntimeException($label . ' expected ' . $expected . ' but got ' . $actual);
    }
}

function assert_contains(string $haystack, string $needle, string $label): void
{
    if (!str_contains($haystack, $needle)) {
        throw new RuntimeException($label . ' should contain ' . $needle . ' but got ' . $haystack);
    }
}

function request_json(string $url): array
{
    $lastError = null;

    for ($attempt = 0; $attempt < 20; $attempt += 1) {
        $http_response_header = [];
        $context = stream_context_create([
            'http' => [
                'ignore_errors' => true,
                'timeout' => 2,
            ],
        ]);

        $body = @file_get_contents($url, false, $context);
        if ($body !== false && isset($http_response_header[0])) {
            preg_match('/\s(\d{3})\s/', $http_response_header[0], $matches);
            $status = isset($matches[1]) ? (int) $matches[1] : 0;
            $headers = [];
            foreach ($http_response_header as $headerLine) {
                if (str_contains($headerLine, ':')) {
                    [$name, $value] = explode(':', $headerLine, 2);
                    $headers[strtolower(trim($name))] = trim($value);
                }
            }
            return [
                'status' => $status,
                'headers' => $headers,
                'bodyRaw' => $body,
                'body' => json_decode($body, true, flags: JSON_THROW_ON_ERROR),
            ];
        }

        $lastError = error_get_last();
        usleep(250000);
    }

    throw new RuntimeException('API indisponible sur ' . $url . ': ' . var_export($lastError, true));
}

function test_calculator_operations(): void
{
    $calculator = new Calculator();
    assert_float_equals(5, $calculator->add(2, 3), 'add');
    assert_float_equals(-1, $calculator->subtract(2, 3), 'subtract');
    assert_float_equals(6, $calculator->multiply(2, 3), 'multiply');
    assert_float_equals(2.5, $calculator->divide(5, 2), 'divide');
}

function test_calculator_rejects_division_by_zero(): void
{
    $calculator = new Calculator();
    try {
        $calculator->divide(10, 0);
        throw new RuntimeException('divide should reject zero division');
    } catch (InvalidArgumentException $error) {
        assert_contains($error->getMessage(), 'Division par zéro impossible.', 'zero division message');
    }
}

function test_http_contract_when_target_is_configured(): void
{
    $baseUrl = getenv('TEST_BASE_URL');
    if ($baseUrl === false || trim($baseUrl) === '') {
        echo "Tests HTTP ignorés : TEST_BASE_URL absent\n";
        return;
    }

    $root = request_json($baseUrl . '/');
    assert_equals(200, $root['status'], 'root status');
    assert_equals(['message' => 'Hello from backend-php'], $root['body'], 'root body');

    $uniqueValue = hrtime(true) % 1000000;
    $add = request_json($baseUrl . '/calculate?operation=add&a=' . $uniqueValue . '&b=3');
    assert_equals(200, $add['status'], 'add status');
    assert_equals('add', $add['body']['operation'], 'add operation');
    assert_equals($uniqueValue + 3, $add['body']['result'], 'add result');
    assert_equals(false, $add['body']['cached'], 'add cache miss');
    assert_equals('application/json; charset=utf-8', $add['headers']['content-type'] ?? '', 'content-type');
    assert_equals('*', $add['headers']['access-control-allow-origin'] ?? '', 'cors header');

    $path = $baseUrl . '/calculate?operation=multiply&a=' . $uniqueValue . '&b=7';
    $firstMultiply = request_json($path);
    $secondMultiply = request_json($path);
    assert_equals(200, $firstMultiply['status'], 'first multiply status');
    assert_equals(200, $secondMultiply['status'], 'second multiply status');
    assert_equals(false, $firstMultiply['body']['cached'], 'first multiply cache miss');
    assert_equals(true, $secondMultiply['body']['cached'], 'second multiply cache hit');

    $divisionByZero = request_json($baseUrl . '/calculate?operation=divide&a=10&b=0');
    assert_equals(400, $divisionByZero['status'], 'division by zero status');
    assert_equals(['error' => 'Division par zéro impossible.'], $divisionByZero['body'], 'division by zero body');

    $unknownOperation = request_json($baseUrl . '/calculate?operation=modulo&a=1&b=2');
    assert_equals(400, $unknownOperation['status'], 'unknown operation status');
    assert_contains($unknownOperation['bodyRaw'], 'Opération inconnue', 'unknown operation body');
}

test_calculator_operations();
test_calculator_rejects_division_by_zero();
test_http_contract_when_target_is_configured();

echo "Tests PHP OK\n";

File: backend-java/tests/TestRunner.java
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

public final class TestRunner {
    private static final HttpClient HTTP_CLIENT = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(2))
        .build();

    private TestRunner() {
    }

    public static void main(String[] args) throws Exception {
        testCalculatorOperations();
        testCalculatorRejectsDivisionByZero();
        testHttpContractWhenTargetIsConfigured();
        System.out.println("Tests Java OK");
    }

    private static void testCalculatorOperations() {
        Calculator calculator = new Calculator();
        assertEquals(5, calculator.add(2, 3), "add");
        assertEquals(-1, calculator.subtract(2, 3), "subtract");
        assertEquals(6, calculator.multiply(2, 3), "multiply");
        assertEquals(2.5, calculator.divide(5, 2), "divide");
    }

    private static void testCalculatorRejectsDivisionByZero() {
        Calculator calculator = new Calculator();
        try {
            calculator.divide(10, 0);
            throw new AssertionError("divide should reject zero division");
        } catch (IllegalArgumentException error) {
            assertContains(error.getMessage(), "Division par zéro impossible.", "zero division message");
        }
    }

    private static void testHttpContractWhenTargetIsConfigured() throws Exception {
        String baseUrl = System.getenv("TEST_BASE_URL");
        if (baseUrl == null || baseUrl.isBlank()) {
            System.out.println("Tests HTTP ignorés : TEST_BASE_URL absent");
            return;
        }

        HttpResult root = request(baseUrl + "/");
        assertEquals(200, root.statusCode, "root status");
        assertContains(root.body, "Hello from backend-java", "root body");

        long uniqueValue = System.nanoTime() % 1_000_000;
        String addPath = baseUrl + "/calculate?operation=add&a=" + uniqueValue + "&b=3";
        HttpResult add = request(addPath);
        assertEquals(200, add.statusCode, "add status");
        assertContains(add.body, "\"operation\":\"add\"", "add operation");
        assertContains(add.body, "\"result\":" + (uniqueValue + 3), "add result");
        assertContains(add.body, "\"cached\":false", "add cache miss");
        assertEquals("application/json; charset=utf-8", add.contentType, "content-type");
        assertEquals("*", add.corsHeader, "cors header");

        String multiplyPath = baseUrl + "/calculate?operation=multiply&a=" + uniqueValue + "&b=7";
        HttpResult firstMultiply = request(multiplyPath);
        HttpResult secondMultiply = request(multiplyPath);
        assertEquals(200, firstMultiply.statusCode, "first multiply status");
        assertEquals(200, secondMultiply.statusCode, "second multiply status");
        assertContains(firstMultiply.body, "\"cached\":false", "first multiply cache miss");
        assertContains(secondMultiply.body, "\"cached\":true", "second multiply cache hit");

        HttpResult divisionByZero = request(baseUrl + "/calculate?operation=divide&a=10&b=0");
        assertEquals(400, divisionByZero.statusCode, "division by zero status");
        assertContains(divisionByZero.body, "Division par zéro impossible.", "division by zero body");

        HttpResult unknownOperation = request(baseUrl + "/calculate?operation=modulo&a=1&b=2");
        assertEquals(400, unknownOperation.statusCode, "unknown operation status");
        assertContains(unknownOperation.body, "Opération inconnue", "unknown operation body");
    }

    private static HttpResult request(String url) throws Exception {
        Exception lastError = null;
        for (int attempt = 0; attempt < 20; attempt += 1) {
            try {
                HttpRequest request = HttpRequest.newBuilder(URI.create(url))
                    .timeout(Duration.ofSeconds(2))
                    .GET()
                    .build();
                HttpResponse<String> response = HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofString());
                return new HttpResult(
                    response.statusCode(),
                    response.body(),
                    response.headers().firstValue("content-type").orElse(""),
                    response.headers().firstValue("access-control-allow-origin").orElse("")
                );
            } catch (IOException | InterruptedException error) {
                lastError = error;
                Thread.sleep(250);
            }
        }
        throw new AssertionError("API indisponible: " + lastError);
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + " expected " + expected + " but got " + actual);
        }
    }

    private static void assertEquals(double expected, double actual, String label) {
        if (Math.abs(expected - actual) > 0.000001) {
            throw new AssertionError(label + " expected " + expected + " but got " + actual);
        }
    }

    private static void assertContains(String actual, String expectedPart, String label) {
        if (!actual.contains(expectedPart)) {
            throw new AssertionError(label + " should contain " + expectedPart + " but got " + actual);
        }
    }

    private record HttpResult(int statusCode, String body, String contentType, String corsHeader) {
    }
}
